x-pack/plugins/observability_solution/entity_manager/server/routes/entities/update.ts

```diff
@@ -0,0 +1,131 @@
+ *         required: false
+ *         schema:
+ *           type: boolean
+ *           default: false
+ *     requestBody:
+ *       description: The definition properties to update
+ *       required: true
+ *       content:
+ *         application/json:
+ *           schema:
+ *             $ref: '#/components/schemas/entityDefinitionUpdateSchema'
+ *     responses:
+ *       200:
+ *         description: Success
+ *         content:
+ *           application/json:
+ *             schema:
+ *               $ref: '#/components/schemas/entityDefinitionSchema'
+ *       400:
+ *         description: The entity definition cannot be installed; see the error for more details
+ *       404:
+ *         description: The entity definition does not exist
+ *       403:
+ *         description: User is not allowed to update the entity definition
+ *       409:
+ *         description: The entity definition is being updated by another request
+ */
+export function updateEntityDefinitionRoute<T extends RequestHandlerContext>({
+  router,
+  server,
+}: SetupRouteOptions<T>) {
+  router.put<{ id: string }, CreateEntityDefinitionQuery, EntityDefinitionUpdate>(
+    {
+      path: '/internal/entities/definition/{id}',
+      validate: {
+        body: entityDefinitionUpdateSchema.strict(),
+        query: createEntityDefinitionQuerySchema,
+        params: z.object({ id: z.string() }),
+      },
+    },
+    async (context, req, res) => {
+      const { logger } = server;
+      const core = await context.core;
+      const soClient = core.savedObjects.client;
+      const esClient = core.elasticsearch.client.asCurrentUser;
+
+      try {
+        const installedDefinition = await findEntityDefinitionById({
+          soClient,
+          esClient,
+          id: req.params.id,
+        });
+
+        if (!installedDefinition) {
+          return res.notFound({
+            body: { message: `Entity definition [${req.params.id}] not found` },
+          });
+        }
+
+        if (installedDefinition.managed) {
+          return res.forbidden({
+            body: { message: `Managed definition can't be updated` },
+          });
+        }
+
+        if (installationInProgress(installedDefinition)) {
+          return res.conflict({
+            body: { message: `Entity definition [${req.params.id}] is being installed` },
+          });
+        }
+
+        const updatedDefinition = await reinstallEntityDefinition({
+          soClient,
+          esClient,
+          logger,
+          definition: installedDefinition,
+          definitionUpdate: req.body,
+        });
+
+        if (!req.query.installOnly) {
+          await startTransform(esClient, updatedDefinition, logger);
+        }
+
+        return res.ok({ body: updatedDefinition });
+      } catch (e) {
+        logger.error(e);
+
+        if (e instanceof EntitySecurityException || e instanceof InvalidTransformError) {
+          return res.customError({ body: e, statusCode: 400 });
```

Comment:
Feels like `EntitySecurityException` should result in a 403, and `InvalidTransformError` does seem like a 400 but could we give the error more context? I'm not sure what the error coming from ES looks like but it would be great if what came out was like "Invalid entity definition changes, the resulting transform is not valid".

Even more ideally, we wouldn't mention transforms at all, but rather express what is wrong in the definition itself.
This could be a follow up thing though.

Reply:
sounds like a good follow up